imap_processing/ultra/l2/ultra_l2.py

```diff
@@ -0,0 +1,483 @@
+        if (
+            not l1c_product.elevation_bin_center.values.shape
+            == l1c_products[0].elevation_bin_center.values.shape
+        ):
+            raise ValueError("Elevation bin center shape mismatch.")
+        if (
+            not l1c_product.energy_bin_center.values.shape
+            == l1c_products[0].energy_bin_center.values.shape
+        ):
+            raise ValueError("Energy bin center shape mismatch.")
+
+    (num_el, num_az, num_energy) = l1c_products[0].counts.shape
+
+    # This will contain the counts for each energy bin
+    combined_counts = np.zeros((num_el * num_az, num_energy))
+
+    # Build the azimuth and elevation grid in the heliocentric ecliptic frame (HAE)
+    (az_range, el_range, az_grid, el_grid) = l2_utils.build_az_el_grid(
+        spacing=spacing_deg,
+        input_degrees=True,
+        output_degrees=False,
+        centered_azimuth=False,  # (0, 2pi rad = 0, 360 deg)
+        centered_elevation=True,  # (-pi/2, pi/2 rad = -90, 90 deg)
+    )
+    # Create cartesian (x,y,z) vectors corresponding to the
+    # grid points in the helio inertial frame
+    radii_helio = geometry.spherical_to_cartesian(
+        np.stack((np.ones_like(az_grid), az_grid, el_grid), axis=-1)
+    )
+    # Unravel to [n, 3] shape
+    radii_helio = radii_helio.reshape(-1, 3, order="C")
+
+    solid_angle_grid = l2_utils.build_solid_angle_map(
+        spacing=spacing_deg, input_degrees=True, output_degrees=False
+    )
+
+    frame_epochs = np.unique([l1c_product.epoch for l1c_product in l1c_products])
+    if len(frame_epochs) != len(l1c_products):
+        logger.warning(
+            "Frame epochs are not unique for each l1c product: "
+            f"{len(frame_epochs)} epochs vs {len(l1c_products)} products."
+        )
+
+    (combined_exptime_45, combined_exptime_90, combined_exptime_total) = (
+        build_dps_combined_exposure_time(l1c_products, spacing_deg)
+    )
+
+    energy_centers_from_file = l1c_products[0].energy_bin_center.values
+    energy_bin_edges, energy_midpoints = build_energy_bins()
+
+    # The function we call adds an extra bin edge at the start of the array.
+    # We remove it here.
+    energy_bin_edges = energy_bin_edges[1:]
+    energy_midpoints = energy_midpoints[1:]
+
+    if not np.allclose(energy_centers_from_file, energy_midpoints):
+        raise ValueError(
+            f"Energy bin centers from file: {energy_centers_from_file}\n"
+            f"Energy midpoints from function: {energy_midpoints}"
+        )
+
+    energy_delta = np.diff(energy_bin_edges)
+    if not np.all(energy_delta >= 0):
+        raise ValueError("Energy bin edges must be monotonically increasing.")
+
+    for prod_num, l1c_prod in enumerate(l1c_products):
+        time = float(l1c_prod.epoch.values)
+        logger.info(
+            f"Projecting exposure time from product at time index {prod_num} "
+            f"with epoch {time} = {spice.et2utc(time, 'ISOC', 3)}"
```

Comment:
Oh... here is the spice call. This should be replaced with a `imap_processing.spice.time.et_to_utc` function that should show up soon (Tori has a PR up with it).